File: apps/api/app/routers/projects.py
"""Projects router — client delivery jobs within a workspace."""

import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.dependencies import CurrentUser
from app.schemas.projects import ProjectCreate, ProjectResponse, ProjectUpdate
from app.services import projects as project_service

router = APIRouter()


def _require_workspace(user: CurrentUser) -> uuid.UUID:
    if user.workspace_id is None:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="User does not belong to a workspace",
        )
    return user.workspace_id


@router.get(
    "/projects",
    response_model=list[ProjectResponse],
    summary="List projects in the current workspace",
)
async def list_projects(
    current_user: CurrentUser,
    db: Annotated[AsyncSession, Depends(get_db)],
) -> list[ProjectResponse]:
    workspace_id = _require_workspace(current_user)
    rows = await project_service.list_projects(db, workspace_id)
    return [ProjectResponse.model_validate(p) for p in rows]


@router.post(
    "/projects",
    response_model=ProjectResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Create a project",
)
async def create_project(
    payload: ProjectCreate,
    current_user: CurrentUser,
    db: Annotated[AsyncSession, Depends(get_db)],
) -> ProjectResponse:
    workspace_id = _require_workspace(current_user)
    project = await project_service.create_project(
        db, workspace_id, current_user.id, payload
    )
    return ProjectResponse.model_validate(project)


@router.get(
    "/projects/{project_id}",
    response_model=ProjectResponse,
    summary="Get a project",
)
async def get_project(
    project_id: uuid.UUID,
    current_user: CurrentUser,
    db: Annotated[AsyncSession, Depends(get_db)],
) -> ProjectResponse:
    workspace_id = _require_workspace(current_user)
    project = await project_service.get_project(db, project_id, workspace_id)
    return ProjectResponse.model_validate(project)


@router.patch(
    "/projects/{project_id}",
    response_model=ProjectResponse,
    summary="Update a project",
)
async def update_project(
    project_id: uuid.UUID,
    payload: ProjectUpdate,
    current_user: CurrentUser,
    db: Annotated[AsyncSession, Depends(get_db)],
) -> ProjectResponse:
    workspace_id = _require_workspace(current_user)
    project = await project_service.update_project(
        db, project_id, workspace_id, payload
    )
    return ProjectResponse.model_validate(project)


@router.delete(
    "/projects/{project_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete a project",
)
async def delete_project(
    project_id: uuid.UUID,
    current_user: CurrentUser,
    db: Annotated[AsyncSession, Depends(get_db)],
) -> None:
    workspace_id = _require_workspace(current_user)
    await project_service.delete_project(db, project_id, workspace_id)

File: apps/mobile/services/projects.ts
/**
 * Projects API service — client delivery jobs in the current workspace.
 */

import { apiDelete, apiGet, apiPatch, apiPost } from "@/services/api";
import type {
  Project,
  ProjectCreatePayload,
  ProjectUpdatePayload,
} from "@/types/projects";

export { normalizeExternalUrl } from "@/services/project-urls";

export async function listProjects(): Promise<Project[]> {
  return apiGet<Project[]>("/api/v1/projects");
}

export async function getProject(id: string): Promise<Project> {
  return apiGet<Project>(`/api/v1/projects/${id}`);
}

export async function createProject(
  payload: ProjectCreatePayload
): Promise<Project> {
  return apiPost<Project>("/api/v1/projects", payload);
}

export async function updateProject(
  id: string,
  payload: ProjectUpdatePayload
): Promise<Project> {
  return apiPatch<Project>(`/api/v1/projects/${id}`, payload);
}

export async function deleteProject(id: string): Promise<void> {
  return apiDelete(`/api/v1/projects/${id}`);
}
